File: composeApp/src/commonMain/kotlin/com/aus/notelikeus/domain/query/SearchText.kt
package com.aus.notelikeus.domain.query

/**
 * Normalisation for search: case folded, diacritics stripped, punctuation reduced to spaces.
 *
 * Search used to be `String.contains(query, ignoreCase = true)` over every note in memory. That
 * matches mid-word, which sounds generous and mostly is not — searching `ote` found "notes" while
 * searching `cafe` did not find "café", because case folding is the only normalisation `contains`
 * does. Anyone typing an accented language got the worse half of both behaviours.
 *
 * The folding table is written out rather than delegated to `java.text.Normalizer`, which is the
 * obvious tool and unavailable here: this is `commonMain`, and while both current targets happen
 * to be JVM there is no shared `jvmMain` source set to hang an `actual` on. Adding one is a build
 * change, and `expect`/`actual` would mean the identical implementation written twice, in two
 * source sets, free to drift. Sixty lines of table has no such failure mode and is directly
 * testable.
 *
 * Covers Latin-1 Supplement and Latin Extended-A, which is every accented form the Latin-script
 * languages this app is likely to hold actually use. Anything outside that range passes through
 * unchanged: a Greek or Cyrillic note is still searchable by its own characters, it simply gets no
 * folding. That is the correct failure — dropping unknown characters would make those notes
 * unfindable rather than merely unfolded.
 */
private fun foldChar(c: Char): String = when (c.code) {
    in 0xC0..0xC5, in 0xE0..0xE5 -> "a"
    0xC6, 0xE6 -> "ae"
    0xC7, 0xE7 -> "c"
    in 0xC8..0xCB, in 0xE8..0xEB -> "e"
    in 0xCC..0xCF, in 0xEC..0xEF -> "i"
    0xD0, 0xF0 -> "d"
    0xD1, 0xF1 -> "n"
    in 0xD2..0xD6, in 0xF2..0xF6, 0xD8, 0xF8 -> "o"
    in 0xD9..0xDC, in 0xF9..0xFC -> "u"
    0xDD, 0xFD, 0xFF -> "y"
    0xDE, 0xFE -> "th"
    0xDF -> "ss"

    // Latin Extended-A
    in 0x100..0x105 -> "a"
    in 0x106..0x10D -> "c"
    in 0x10E..0x111 -> "d"
    in 0x112..0x11B -> "e"
    in 0x11C..0x123 -> "g"
    in 0x124..0x127 -> "h"
    in 0x128..0x131 -> "i"
    0x132, 0x133 -> "ij"
    0x134, 0x135 -> "j"
    in 0x136..0x138 -> "k"
    in 0x139..0x142 -> "l"
    in 0x143..0x14B -> "n"
    in 0x14C..0x151 -> "o"
    0x152, 0x153 -> "oe"
    in 0x154..0x159 -> "r"
    in 0x15A..0x161 -> "s"
    in 0x162..0x167 -> "t"
    in 0x168..0x173 -> "u"
    0x174, 0x175 -> "w"
    in 0x176..0x178 -> "y"
    in 0x179..0x17E -> "z"
    0x17F -> "s"

    else -> c.toString()
}

/**
 * The searchable form of a string: lower case, folded, with every non-alphanumeric run collapsed
 * to a single space and the ends trimmed.
 *
 * Punctuation becomes a separator rather than being deleted, so `don't` tokenises as `don` + `t`
 * rather than `dont`. That is deliberate: it means a search for `don` finds it, and the
 * alternative — gluing the halves — invents a word the note does not contain.
 */
fun foldForSearch(input: String): String {
    val out = StringBuilder(input.length)
    var pendingSpace = false
    for (raw in input) {
        val lowered = raw.lowercaseChar()
        val folded = foldChar(lowered)
        val isWord = folded.all { it.isLetterOrDigit() }
        if (folded.isEmpty() || !isWord) {
            // Collapse separators, and never emit a leading one.
            if (out.isNotEmpty()) pendingSpace = true
            continue
        }
        if (pendingSpace) {
            out.append(' ')
            pendingSpace = false
        }
        out.append(folded)
    }
    return out.toString()
}

/** The folded tokens of [input], in order, with duplicates preserved. */
fun searchTokens(input: String): List<String> =
    foldForSearch(input).split(' ').filter { it.isNotEmpty() }

/**
 * The value stored in `notes.searchText`.
 *
 * Everything the old in-memory search looked at — title, body, checklist item text and label
 * names — flattened into one folded string, so a single indexed column answers what previously
 * required four passes over every note plus its joined rows.
 *
 * Label *names* are included rather than ids because that is what a user types. It also means the
 * column goes stale when a label is renamed, which the repository handles by rewriting every note
 * carrying that label — the same set it already rewrites to keep the cloud copies current.
 */
fun buildSearchText(
    title: String,
    content: String,
    checklistTexts: List<String>,
    labelNames: List<String>
): String {
    val parts = buildList {
        add(title)
        add(content)
        addAll(checklistTexts)
        addAll(labelNames)
    }
    return foldForSearch(parts.filter { it.isNotBlank() }.joinToString(" "))
}

File: web/src/lib/text/searchText.ts
/**
 * Search normalisation matching Kotlin `domain.query.SearchText`.
 *
 * Token-prefix + diacritic fold, not mid-word `includes()`. "ote" must not find "notes";
 * "cafe" must find "café". The fold table is copied from commonMain rather than using
 * String.normalize, so a Latin-1 / Extended-A note matches the same way on every client.
 */

function foldChar(code: number): string {
  if ((code >= 0xc0 && code <= 0xc5) || (code >= 0xe0 && code <= 0xe5)) return 'a';
  if (code === 0xc6 || code === 0xe6) return 'ae';
  if (code === 0xc7 || code === 0xe7) return 'c';
  if ((code >= 0xc8 && code <= 0xcb) || (code >= 0xe8 && code <= 0xeb)) return 'e';
  if ((code >= 0xcc && code <= 0xcf) || (code >= 0xec && code <= 0xef)) return 'i';
  if (code === 0xd0 || code === 0xf0) return 'd';
  if (code === 0xd1 || code === 0xf1) return 'n';
  if (
    (code >= 0xd2 && code <= 0xd6) ||
    (code >= 0xf2 && code <= 0xf6) ||
    code === 0xd8 ||
    code === 0xf8
  ) {
    return 'o';
  }
  if ((code >= 0xd9 && code <= 0xdc) || (code >= 0xf9 && code <= 0xfc)) return 'u';
  if (code === 0xdd || code === 0xfd || code === 0xff) return 'y';
  if (code === 0xde || code === 0xfe) return 'th';
  if (code === 0xdf) return 'ss';

  if (code >= 0x100 && code <= 0x105) return 'a';
  if (code >= 0x106 && code <= 0x10d) return 'c';
  if (code >= 0x10e && code <= 0x111) return 'd';
  if (code >= 0x112 && code <= 0x11b) return 'e';
  if (code >= 0x11c && code <= 0x123) return 'g';
  if (code >= 0x124 && code <= 0x127) return 'h';
  if (code >= 0x128 && code <= 0x131) return 'i';
  if (code === 0x132 || code === 0x133) return 'ij';
  if (code === 0x134 || code === 0x135) return 'j';
  if (code >= 0x136 && code <= 0x138) return 'k';
  if (code >= 0x139 && code <= 0x142) return 'l';
  if (code >= 0x143 && code <= 0x14b) return 'n';
  if (code >= 0x14c && code <= 0x151) return 'o';
  if (code === 0x152 || code === 0x153) return 'oe';
  if (code >= 0x154 && code <= 0x159) return 'r';
  if (code >= 0x15a && code <= 0x161) return 's';
  if (code >= 0x162 && code <= 0x167) return 't';
  if (code >= 0x168 && code <= 0x173) return 'u';
  if (code === 0x174 || code === 0x175) return 'w';
  if (code >= 0x176 && code <= 0x178) return 'y';
  if (code >= 0x179 && code <= 0x17e) return 'z';
  if (code === 0x17f) return 's';

  return String.fromCodePoint(code).toLowerCase();
}

function isLetterOrDigit(ch: string): boolean {
  return /\p{L}|\p{N}/u.test(ch);
}

/** Lower case, diacritics stripped, punctuation collapsed to single spaces. */
export function foldForSearch(input: string): string {
  let out = '';
  let pendingSpace = false;
  for (const raw of input) {
    const folded = foldChar(raw.codePointAt(0) ?? 0);
    const isWord = folded.length > 0 && [...folded].every(isLetterOrDigit);
    if (folded.length === 0 || !isWord) {
      if (out.length > 0) pendingSpace = true;
      continue;
    }
    if (pendingSpace) {
      out += ' ';
      pendingSpace = false;
    }
    out += folded;
  }
  return out;
}

export function searchTokens(input: string): string[] {
  return foldForSearch(input).split(' ').filter((token) => token.length > 0);
}

export function buildSearchText(
  title: string,
  content: string,
  checklistTexts: string[],
  labelNames: string[],
): string {
  const parts = [title, content, ...checklistTexts, ...labelNames].filter((part) => part.trim());
  return foldForSearch(parts.join(' '));
}

/**
 * Every query token must be a word-prefix somewhere in the note, in any order.
 * Mirrors Kotlin `NoteQueryMatcher.matchesText`.
 */
export function noteMatchesSearchQuery(
  haystack: string,
  query: string,
): boolean {
  const needles = searchTokens(query);
  if (needles.length === 0) return true;
  return needles.every((token) => haystack.startsWith(token) || haystack.includes(` ${token}`));
}
